src/kafka_endpoint.cc

```diff
@@ -0,0 +1,971 @@
+        {{"bootstrap.servers", port_}, {"enable.auto.commit", "true"}});
+
+    // Create a consumer instance
+    consumer_ = std::make_unique<kafka::clients::KafkaConsumer>(props);
+
+    // Subscribe to topics
+    // TODO: Currently if a single topic in a list is valid, the server
+    // will not time out, it will silently fail.
+    consumer_->subscribe({consumer_topics_});
+  }
+  catch (const kafka::KafkaException& e) {
+    return TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_INTERNAL,
+        (e.what() + std::string(" Check topics exist.")).c_str());
+  }
+  consumer_active_ = true;
+  consumer_thread_ = std::thread(&KafkaEndpoint::ConsumeRequests, this);
+
+  return nullptr;  // Success
+}
+
+void
+KafkaEndpoint::CreateInferenceResponse(
+    std::vector<std::pair<std::string, std::string>>& header_pair_vector,
+    const std::string& val, const std::string& request_id)
+{
+  std::string response_topic =
+      current_request_map_[request_id]["response_topic"];
+  kafka::Key key = kafka::NullKey;
+  // TODO: This functionality needs to be tested
+  if (!current_request_map_[request_id]["response_key"].empty()) {
+    kafka::Value key_string(
+        current_request_map_[request_id]["response_key"].c_str(),
+        current_request_map_[request_id]["response_key"].size());
+    key = kafka::Key(key_string);
+  }
+  LOG_INFO << "Response Topic: " << response_topic;
+  LOG_INFO << "Creating inference response";
+  std::unique_ptr<kafka::clients::producer::ProducerRecord> response_record(
+      std::make_unique<kafka::clients::producer::ProducerRecord>(
+          response_topic, key, kafka::Value(val.c_str(), val.size())));
+  LOG_INFO << "[" << &response_record << "]";
+
+  for (auto it = header_pair_vector.begin(); it != header_pair_vector.end();
+       ++it) {
+    response_record->headers().emplace_back(
+        it->first, kafka::Header::Value(it->second.c_str(), it->second.size()));
+  }
+  ProduceInferenceResponse(response_record);
+  // Release the request ID so that it can used again
+  current_request_map_.erase(request_id);
+}
+
+
+TRITONSERVER_Error*
+KafkaEndpoint::ProduceInferenceResponse(
+    std::unique_ptr<kafka::clients::producer::ProducerRecord>&
+        producer_response_msg)
+{
+  try {
+    LOG_INFO << "Response Sending! "
+             << producer_response_msg->value().toString();
+    kafka::clients::producer::RecordMetadata metadata =
+        producer_->syncSend(*producer_response_msg);
+    LOG_INFO << "Response Sent! " << producer_response_msg->value().toString();
+  }
+  catch (const kafka::KafkaException& e) {
+    return TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_INTERNAL,
+        ("Message delivery failed: " + e.error().message()).c_str());
+  }
+  return nullptr;  // Success
+}
+
+void
+KafkaEndpoint::ConsumeRequests()
+{
+  try {
+    DisplayConsumerTopics();
+
+    while (consumer_active_) {
+      auto records = consumer_->poll(std::chrono::milliseconds(100));
```

Comment:
Is there an event based system where consumer wakes up when there's a request, or does it have to be polled? Is the 100ms arbitrary or a best practice?

Reply:
Should add this to the doc